compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -1101,20 +1101,88 @@ private STNode parseFuncDefOrFuncTypeDesc(STNode metadata, STNode visibilityQual
+     * <p>
+     * <code>
+     * function-defn := `function` identifier function-signature function-body
+     * <br/>
+     * function-type-descriptor := `function`
+     * </code>
+     *
+     * @param metadata Preceding metadata
+     * @param visibilityQualifier Preceding visibility qualifier
+     * @param qualifiers Preceding visibility qualifier
+     * @param functionKeyword `function` keyword
+     * @param isObjectMember Whether object member or not
+     * @param isObjectTypeDesc Whether object type or not
+     * @return Parsed node
+     */
+    private STNode parseFuncDefOrFuncTypeDesc(STNode metadata, STNode visibilityQualifier,
+                                              List<STNode> qualifiers, STNode functionKeyword, boolean isObjectMember,
+                                              boolean isObjectTypeDesc) {
+        switch (peek(2).kind) {
+            case OPEN_PAREN_TOKEN:
                 STNode name = parseFunctionName();
-                STNode resourcePath = parseOptionalRelativePath(isObjectMember);
+                STNode resourcePath = STNodeFactory.createEmptyNodeList();
                 switchContext(ParserRuleContext.FUNC_DEF);
                 STNode funcSignature = parseFuncSignature(false);
                 STNode funcDef = parseFuncDefOrMethodDeclEnd(metadata, visibilityQualifier, qualifiers, functionKeyword,
                         name, resourcePath, funcSignature, isObjectMember, isObjectTypeDesc);
                 endContext();
                 return funcDef;
+            case DOT_TOKEN:
+            case IDENTIFIER_TOKEN:
+            case OPEN_BRACKET_TOKEN:
+                name = parseFunctionName();
+                resourcePath = parseRelativeResourcePath();
+                if (!isObjectMember) {
+                    addInvalidNodeToNextToken(resourcePath,
+                                              DiagnosticErrorCode.ERROR_RESOURCE_PATH_IN_FUNCTION_DEFINITION);
+                    resourcePath = STNodeFactory.createEmptyNodeList();
+                }
+                switchContext(ParserRuleContext.FUNC_DEF);
+                funcSignature = parseFuncSignature(false);
+                funcDef = parseFuncDefOrMethodDeclEnd(metadata, visibilityQualifier, qualifiers, functionKeyword,
+                        name, resourcePath, funcSignature, isObjectMember, isObjectTypeDesc);
+                endContext();
+                return funcDef;
+            case EQUAL_TOKEN:
+            case SEMICOLON_TOKEN:
+                endContext();
+                return parseVarDeclWithFunctionType(metadata, visibilityQualifier, qualifiers, functionKeyword,
+                                                    STNodeFactory.createEmptyNode(), isObjectMember, isObjectTypeDesc,
+                                                    false);
+            default:
+                STToken functionName = consume();
+                STToken token = peek();
```

Comment:
Lest use `STToken nextNexttoken = peek(2);`  before the switch and use it here and  in switch case

Reply:
there's a separate method to get the next-next-token. `getNextNextToken()` when doing so. But I think using a `peek()` in the recovery() call is better to avoid mistakes.

Reply:
No right? any example?

Reply:
As discussed offline let's keep this as it is and directly peek inside recovery without using new variable